clustered/client/src/main/java/org/ehcache/clustered/client/internal/config/xml/ClusteringCacheManagerServiceConfigurationParser.java

```diff
@@ -0,0 +1,466 @@
+    if (clusteringServiceConfiguration.getConnectionSource() instanceof ConnectionSource.ClusterUri) {
+      return createRootUrlElement(doc, clusteringServiceConfiguration);
+    }
+    return createServerElement(doc, clusteringServiceConfiguration);
+  }
+
+  private void processTimeUnits(Document doc, Element parent, ClusteringServiceConfiguration clusteringServiceConfiguration) {
+    if (clusteringServiceConfiguration.getTimeouts() != null) {
+      Timeouts timeouts = clusteringServiceConfiguration.getTimeouts();
+
+      Element readTimeoutElem = createTimeoutElement(doc, READ_TIMEOUT_ELEMENT_NAME, timeouts.getReadOperationTimeout());
+      Element writeTimeoutElem = createTimeoutElement(doc, WRITE_TIMEOUT_ELEMENT_NAME, timeouts.getWriteOperationTimeout());
+      Element connectionTimeoutElem = createTimeoutElement(doc, CONNECTION_TIMEOUT_ELEMENT_NAME, timeouts.getConnectionTimeout());
+      /*
+      Important: do not change the order of following three elements if corresponding change is not done in xsd
+       */
+      parent.appendChild(readTimeoutElem);
+      parent.appendChild(writeTimeoutElem);
+      parent.appendChild(connectionTimeoutElem);
+    }
+  }
+
+  private Element createTimeoutElement(Document doc, String timeoutName, Duration timeout) {
+    Element retElement = null;
+    if (READ_TIMEOUT_ELEMENT_NAME.equals(timeoutName)) {
+      retElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + READ_TIMEOUT_ELEMENT_NAME);
+    } else if (WRITE_TIMEOUT_ELEMENT_NAME.equals(timeoutName)) {
+      retElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + WRITE_TIMEOUT_ELEMENT_NAME);
+    } else {
+      retElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + CONNECTION_TIMEOUT_ELEMENT_NAME);
+    }
+    retElement.setAttribute(UNIT_ATTRIBUTE_NAME, DEFAULT_UNIT_ATTRIBUTE_VALUE);
+    retElement.setTextContent(Long.toString(timeout.getSeconds()));
+    return retElement;
+  }
+
+  private Element processServerSideElements(Document doc, ClusteringServiceConfiguration clusteringServiceConfiguration) {
+    Element serverSideConfigurationElem = createServerSideConfigurationElement(doc, clusteringServiceConfiguration);
+
+    if (clusteringServiceConfiguration.getServerConfiguration() != null) {
+      ServerSideConfiguration serverSideConfiguration = clusteringServiceConfiguration.getServerConfiguration();
+      String defaultServerResource = serverSideConfiguration.getDefaultServerResource();
+      if (!(defaultServerResource == null || defaultServerResource.trim().length() == 0)) {
+        Element defaultResourceElement = createDefaultServerResourceElement(doc, defaultServerResource);
+        serverSideConfigurationElem.appendChild(defaultResourceElement);
+      }
+      Map<String, ServerSideConfiguration.Pool> resourcePools = serverSideConfiguration.getResourcePools();
+      if (!(resourcePools == null || resourcePools.isEmpty())) {
```

Comment:
No need to call `isEmpty()` the forEach works just as well on an empty collection. Also, is it legal for getResourcePools() to return null?  If not why allow it - just let this code blow up with an NPE.